datalad_next/utils/credman.py

```diff
@@ -553,6 +558,116 @@ def get_sort_key(x):
+        falls back onto querying for a credential matching a set of specified
+        properties (as key-value mappings). If no suitable credential is
+        known, a user is prompted to enter one interactively (if possible
+        in the current session).
+
+        If a credential was entered manually, any given ``type_hint`` will
+        be included as a ``type`` property of the returned credential.
+        Likewise, any ``realm`` property included in the ``query_props``
+        is included in the returned credential in this case.
+
+        If desired, a credential workflow can be completed, after a credential
+        was found to be valid/working, by storing or updating it in the
+        credential store::
+
+            cm = CredentialManager()
+            cname, cprops = cm.obtain(...)
+            # verify credential is working
+            ...
+            # set/update
+            cm.set(cname, _lastused=True, **cprops)
+
+        In the code sketch above, if ``cname`` is ``None`` (as it will be for
+        a newly entered credential, ``set()`` will prompt for a name to store
+        the credential under, and will offer a user the choice to skip storing
+        a credential. For any previously known credential, the ``last-used``
+        property will be updated to enable preferred selection in future
+        credential discovery attempts via ``obtain()``.
+
+        Parameters
+        ----------
+        name: str, optional
+          Name of the credential to be retrieved
+        prompt: str, optional
+          Passed to ``CredentialManager.get()`` if a credential name was
+          provided, or no suitable credential could be found by querying.
+        type_hint: str, optional
+          In case no ``type`` property is included in ``query_props``,
+          this parameter is passed to ``CredentialManager.get()``.
+        query_props: dict, optional
+          Credential property to be used for querying for a suitable
+          credential. When multiple credentials match a query, the last-used
+          credential is selected.
+        expected_props: list or tuple, optional
+          When specified, a credential will be inspected to contain properties
+          matching all listed property names, or a ``ValueError`` will be
+          raised.
+
+        Returns
+        -------
+        (str, dict)
+          Credential name (possibly different from the input, when a credential
+          was discovered based on properties), and credential properties.
+
+        Raises
+        ------
+        ValueError
+          Raised when no matching credential could be found and none was
+          entered. Also raised, when a credential selected from a query result
+          or a manually entered one is missing any of the properties with a
+          name given in ``expected_props``.
+        """
+        cred = None
+        if not name:
+            if query_props:
```

Comment:
Sorry if I'm missing something, but the docstring (first sentence) sounds as if I should be able to obtain a credential solely by prompt, without the need for query parameters. This conditional would prevent this, if I see it correctly

Reply:
Yes, that observation is correct. The type hint is needed such that the credential manager knows what to prompt for (some credentials need more than just a secret). This limitation could be lifted (to be able to just ask for a secret). Or we add docs, as you suggested.

I am always hesitant re examples. Unless executable (which most of the time required added/undesired complexity) they become invalid quickly, which often goes unnoticed. But I do see the issue.

Reply:
I have pushed some examples. Would that work for you?

Reply:
Thx